File: mcp_server/session_store.py
"""Server-side session carts for mock MCP (multi-turn state).

Sessions are kept in an in-process dict with LRU eviction once the pool
exceeds MAX_SESSIONS, so the server doesn't leak memory indefinitely.
"""

from __future__ import annotations

import time
from collections import OrderedDict
from dataclasses import dataclass, field
from typing import Any

_DEFAULT_SESSION = "default_mock_session"
MAX_SESSIONS = 200  # evict oldest when exceeded


@dataclass
class FoodCart:
    restaurant_id: str = ""
    restaurant_name: str = ""
    address_id: str = ""
    lines: list[dict[str, Any]] = field(default_factory=list)
    coupon_code: str | None = None
    coupon_discount_inr: int = 0


@dataclass
class ImCart:
    selected_address_id: str = ""
    lines: list[dict[str, Any]] = field(default_factory=list)


@dataclass
class SessionState:
    food: FoodCart = field(default_factory=FoodCart)
    im: ImCart = field(default_factory=ImCart)
    applied_coupon: str | None = None
    created_at: float = field(default_factory=time.time)
    last_accessed: float = field(default_factory=time.time)


# Use an OrderedDict so we can do O(1) LRU eviction
_sessions: OrderedDict[str, SessionState] = OrderedDict()


def _evict_lru() -> None:
    """Remove the least-recently-used session if the pool is full."""
    while len(_sessions) >= MAX_SESSIONS:
        _sessions.popitem(last=False)  # FIFO = oldest first


def resolve_session_id(params: dict[str, Any] | None) -> str:
    if not params:
        return _DEFAULT_SESSION
    for key in ("requestId", "request_id", "sessionId", "session_id"):
        val = params.get(key)
        if val and str(val).strip():
            return str(val).strip()
    return _DEFAULT_SESSION


def get_session(session_id: str | None = None, params: dict[str, Any] | None = None) -> SessionState:
    sid = session_id or resolve_session_id(params)
    if sid in _sessions:
        session = _sessions[sid]
        session.last_accessed = time.time()
        # Move to end (most recently used)
        _sessions.move_to_end(sid)
        return session
    # New session — evict if needed
    _evict_lru()
    state = SessionState()
    _sessions[sid] = state
    return state


def clear_food_cart(session_id: str) -> None:
    s = get_session(session_id)
    s.food = FoodCart()


def clear_im_cart(session_id: str) -> None:
    s = get_session(session_id)
    s.im = ImCart()


def list_sessions() -> list[dict[str, Any]]:
    """Return debug info for all active sessions."""
    return [
        {
            "session_id": sid,
            "food_items": len(s.food.lines),
            "im_items": len(s.im.lines),
            "created_at": s.created_at,
            "last_accessed": s.last_accessed,
        }
        for sid, s in _sessions.items()
    ]


def session_count() -> int:
    return len(_sessions)

File: frontend/lib/mcp-mock/session-store.ts
type FoodCart = {
  restaurant_id: string;
  restaurant_name: string;
  address_id: string;
  lines: { item_id: string; name: string; qty: number; unit_price_inr: number; line_total_inr: number }[];
  coupon_code?: string;
  coupon_discount?: number;
};

type ImCart = {
  address_id: string;
  lines: { spinId: string; product_id: string; name: string; qty: number; unit_price_inr: number; line_total_inr: number }[];
};

type Session = {
  food: FoodCart;
  im: ImCart;
  food_orders: Record<string, unknown>[];
  im_orders: Record<string, unknown>[];
  bookings: Record<string, unknown>[];
  last_food_order_id?: string;
  last_im_order_id?: string;
  last_booking_id?: string;
};

const sessions = new Map<string, Session>();

function emptySession(): Session {
  return {
    food: { restaurant_id: "", restaurant_name: "", address_id: "", lines: [] },
    im: { address_id: "", lines: [] },
    food_orders: [],
    im_orders: [],
    bookings: [],
  };
}

export function getSession(sid: string): Session {
  if (!sessions.has(sid)) sessions.set(sid, emptySession());
  return sessions.get(sid)!;
}

export function resolveSessionId(params: Record<string, unknown>): string {
  const raw = params.requestId ?? params.request_id ?? params.sessionId ?? params.session_id;
  return typeof raw === "string" && raw.trim() ? raw.trim() : "default-session";
}
